File: src/test/resources/PalindromePerm.java
public class Solution {
    /**
     * @param phrase string to be checked if any permutation is palindrome
     * @return  Indicate if string is a permutation of palindrome
     */
    public boolean isPermutationOfPalindrome(String phrase) {
        if (phrase == "" || phrase == null) return false;

        int[] table = buildCharFrequencyTable(phrase.toLowerCase());
        return checkMaxOneOdd(table);
    }

    private boolean checkMaxOneOdd(int[] table) {
        boolean foundOdd = false;

        for (int count : table) {
            if (count % 2 == 1) {
                if (foundOdd) {
                    return false;
                }
                foundOdd = true;
            }
        }

        return true;
    }

    private int getCharNumber(Character c) {
        int a = Character.getNumericValue('a');
        int z = Character.getNumericValue('z');
        int val = Character.getNumericValue(c);

        if (a <= val && val <= z) {
            return val - a;
        }
        return -1;
    }

    private int[] buildCharFrequencyTable(String phrase) {
        int[] table = new int[Character.getNumericValue('z') -
                                Character.getNumericValue('a') + 1];

        for (char c : phrase.toCharArray()) {
            int x = getCharNumber(c);
            if (x != -1) {
                table[x]++;
            }
        }

        return table;
    }
}

File: src/test/resources/PalindromePerm.kt
class PalindromePerm {
    /**
     * @param phrase string to be checked if any permutation is palindrome
     * *
     * @return  Indicate if string is a permutation of palindrome
     */
    fun isPermutationOfPalindrome(phrase: String?): Boolean {
        if (phrase === "" || phrase == null) return false

        val table = buildCharFrequencyTable(phrase.toLowerCase())
        return checkMaxOneOdd(table)
    }

    private fun checkMaxOneOdd(table: IntArray): Boolean {
        var foundOdd = false

        for (count in table) {
            if (count % 2 == 1) {
                if (foundOdd) {
                    return false
                }
                foundOdd = true
            }
        }

        return true
    }

    private fun getCharNumber(c: Char?): Int {
        val a = Character.getNumericValue('a')
        val z = Character.getNumericValue('z')
        val value = Character.getNumericValue(c!!)

        if (a <= value && value <= z) {
            return value - a
        }
        return -1
    }

    private fun buildCharFrequencyTable(phrase: String): IntArray {
        val table = IntArray(Character.getNumericValue('z') - Character.getNumericValue('a') + 1)

        phrase.toCharArray()
                .map { getCharNumber(it) }
                .filter { it != -1 }
                .forEach { table[it]++ }

        return table
    }
}
